authenticatorbridge/src/main/java/com/bitwarden/authenticatorbridge/manager/AuthenticatorBridgeManagerImpl.kt

```diff
@@ -0,0 +1,193 @@
+        }
+    }
+
+    private fun onAccountsSync(data: EncryptedSharedAccountData) {
+        // Received account sync update. Decrypt with local symmetric key and update StateFlow:
+        mutableSharedAccountsStateFlow.value = symmetricKeyStorageProvider.symmetricKey
+            ?.let { data.decrypt(it) }
+            ?.getOrNull()
+            ?.let { AccountSyncState.Success(it.accounts) }
+            ?: AccountSyncState.Error
+    }
+
+    private fun onServiceConnected(service: IBinder) {
+        bridgeService = IAuthenticatorBridgeService.Stub.asInterface(service)
+
+        // TODO: Add check for version mismatch between client and server SDKs: BITAU-72
+
+        // Ensure we are using the correct symmetric key:
+        val localKeyFingerprint =
+            symmetricKeyStorageProvider.symmetricKey?.toFingerprint()?.getOrNull()
+
+        // Query bridge service to see if we have a matching symmetric key:
+        val haveCorrectKey = bridgeService
+            .safeCall { checkSymmetricEncryptionKeyFingerprint(localKeyFingerprint) }
+            .fold(
+                onSuccess = { it },
+                onFailure = { false },
+            )
+            ?: false
+
+        if (!haveCorrectKey) {
+            // If we don't have the correct key, query for key:
+            symmetricKeyStorageProvider.symmetricKey =
+                bridgeService.safeCall { symmetricEncryptionKeyData }.getOrNull()
+        }
+
+        if (symmetricKeyStorageProvider.symmetricKey == null) {
+            // This means bridgeService returned a null key, which means we can make
+            // no valid operations. We should disconnect form the service and expose to the
+            // calling application that authenticator sync is not enabled.
+            mutableSharedAccountsStateFlow.value = AccountSyncState.SyncNotEnabled
+            unbindService()
+            return
+        }
+
+        // Register callback:
+        bridgeService.safeCall { registerBridgeServiceCallback(authenticatorBridgeCallback) }
+
+        // Sync data:
+        bridgeService.safeCall { syncAccounts() }
+    }
+
+    private fun onServiceDisconnected() {
+        bridgeService = null
+    }
+
+    private fun unbindService() {
+        bridgeService.safeCall { unregisterBridgeServiceCallback(authenticatorBridgeCallback) }
+        bridgeService = null
+        try {
+            applicationContext.unbindService(bridgeServiceConnection)
+        } catch (_: Exception) {
+            // We want to be super safe when unbinding to assure no crashes.
+        }
+    }
+}
+
+/**
+ * Helper function for wrapping all calls to [IAuthenticatorBridgeService] around try catch.
+ *
+ * This is important because all calls to [IAuthenticatorBridgeService] can throw
+ * DeadObjectExceptions as well as RemoteExceptions.
+ */
+fun <T> IAuthenticatorBridgeService?.safeCall(action: IAuthenticatorBridgeService.() -> T): Result<T?> =
```

Comment:
This can be private right? Or do you need this for testing? If you need it for testing, we should move it to a different package and name it `internal`